File: old/gameplay-old/behaviors/Bump.cpp
#include "Bump.hpp"
#include <stdio.h>

using namespace Geometry2d;

namespace Gameplay
{
	namespace Behaviors
	{
		REGISTER_CONFIGURABLE(Bump)
	}
}

ConfigBool   *Gameplay::Behaviors::Bump::_face_ball;
ConfigDouble *Gameplay::Behaviors::Bump::_drive_around_dist;
ConfigDouble *Gameplay::Behaviors::Bump::_setup_to_charge_thresh;
ConfigDouble *Gameplay::Behaviors::Bump::_escape_charge_thresh;
ConfigDouble *Gameplay::Behaviors::Bump::_setup_ball_avoid;
ConfigDouble *Gameplay::Behaviors::Bump::_bump_complete_dist;
ConfigDouble *Gameplay::Behaviors::Bump::_accel_bias;
ConfigDouble *Gameplay::Behaviors::Bump::_facing_thresh;

void Gameplay::Behaviors::Bump::createConfiguration(Configuration *cfg)
{
	_drive_around_dist = new ConfigDouble(cfg, "Bump/Drive Around Dist", 0.45);
	_face_ball = new ConfigBool(cfg, "Bump/Face Ball otherwise target", true);
	_setup_to_charge_thresh = new ConfigDouble(cfg, "Bump/Charge Thresh", 0.1);
	_escape_charge_thresh = new ConfigDouble(cfg, "Bump/Escape Charge Thresh", 0.1);
	_setup_ball_avoid = new ConfigDouble(cfg, "Bump/Setup Ball Avoid", 1.0);
	_bump_complete_dist = new ConfigDouble(cfg, "Bump/Bump Complete Distance", 0.5);
	_accel_bias = new ConfigDouble(cfg, "Bump/Accel Bias", 0.1);
	_facing_thresh = new ConfigDouble(cfg, "Bump/Facing Thresh - Deg", 10);
}

Gameplay::Behaviors::Bump::Bump(GameplayModule *gameplay):
    SingleRobotBehavior(gameplay)
{
	_state = State_Setup;
	target = Point(0.0, Field_Length);
}

void Gameplay::Behaviors::Bump::restart()
{
	_state = State_Setup;
}

bool Gameplay::Behaviors::Bump::run()
{
	if (!robot || !robot->visible)
	{
		return false;
	}
	
	Line targetLine(ball().pos, target);
	const Point dir = Point::direction(robot->angle * DegreesToRadians);
	double facing_thresh = cos(*_facing_thresh * DegreesToRadians);
	double facing_err = dir.dot((target - ball().pos).normalized());
//	robot->addText(QString("Err:%1,T:%2").arg(facing_err).arg(facing_thresh));

	// State changes
	if (_state == State_Setup)
	{
		if (targetLine.distTo(robot->pos) <= *_setup_to_charge_thresh &&
				targetLine.delta().dot(robot->pos - ball().pos) <= -Robot_Radius &&
				facing_err >= facing_thresh)
		{
			_state = State_Charge;
		}
	} else if (_state == State_Charge)
	{
		if (Line(robot->pos, target).distTo(ball().pos) > *_escape_charge_thresh)
		{
			// Ball is in a bad place
			_state = State_Setup;
		}

		// FIXME: should finish at some point, but this condition is bad
//		if (!robot->pos.nearPoint(ball().pos, *_bump_complete_dist))
//		{
//			_state = State_Done;
//		}
	}
	
	// Driving
	if (_state == State_Setup)
	{
		// Move onto the line containing the ball and the_setup_ball_avoid target
		robot->addText(QString("%1").arg(targetLine.delta().dot(robot->pos - ball().pos)));
		Segment behind_line(ball().pos - targetLine.delta().normalized() * (*_drive_around_dist + Robot_Radius),
				ball().pos - targetLine.delta().normalized() * 5.0);
		if (targetLine.delta().dot(robot->pos - ball().pos) > -Robot_Radius)
		{
			// We're very close to or in front of the ball
			robot->addText("In front");
			robot->avoidBallRadius(*_setup_ball_avoid);
			robot->move(ball().pos - targetLine.delta().normalized() * (*_drive_around_dist + Robot_Radius));
		} else {
			// We're behind the ball
			robot->addText("Behind");
			robot->avoidBallRadius(*_setup_ball_avoid);
			robot->move(behind_line.nearestPoint(robot->pos));
			state()->drawLine(behind_line);
		}

		// face in a direction so that on impact, we aim at goal
		Point delta_facing = target - ball().pos;
		robot->face(robot->pos + delta_facing);

	} else if (_state == State_Charge)
	{
		robot->addText("Charge!");
		state()->drawLine(robot->pos, target, Qt::white);
		state()->drawLine(ball().pos, target, Qt::white);

		Point ballToTarget = (target - ball().pos).normalized();
//		Point robotToBall = (ball().pos - robot->pos).normalized();
		Point driveDirection = (ball().pos - ballToTarget * Robot_Radius) - robot->pos;
		
		//We want to move in the direction of the target without path planning
		double speed =  robot->vel.mag() + *_accel_bias; // enough of a bias to force it to accelerate
		robot->worldVelocity(driveDirection.normalized() * speed);
		robot->face(ball().pos);
	} else {
		robot->addText("Done");
		return false;
	}
	
	return true;
}

File: soccer/gameplay/skills/bump.py
import single_robot_behavior
import behavior
import constants
import main
import enum
import robocup
import math


## pushes the ball by bumping into it
class Bump(single_robot_behavior.SingleRobotBehavior):

    # tuneable constants
    DriveAroundDist = 0.35  # how far away from the ball we should stay in the lineup state
    LineupBallAvoidRadius = 0.043  # ball avoid radius
    FacingThresh = 10 * constants.DegreesToRadians  # angle in radians that we must be less than away from the ball
    AccelBias = 0.1  # how much to add to our desired speed
    EscapeChargeThresh = 0.1  # if we're off by this much, we go back to the lineup state
    LineupToChargeThresh = 0.05  # how close we have to be to our target line to enter the charge state

    class State(enum.Enum):
        lineup = 1
        charge = 2

    def __init__(self):
        super().__init__(continuous=False)

        self.target = robocup.Point(0, constants.Field.Length)

        self.add_state(Bump.State.lineup, behavior.Behavior.State.running)
        self.add_state(Bump.State.charge, behavior.Behavior.State.running)

        self.add_transition(behavior.Behavior.State.start, Bump.State.lineup,
                            lambda: True, 'immediately')

        self.add_transition(
            Bump.State.lineup, Bump.State.charge,
            lambda: self.target_line().dist_to(self.robot.pos) <= Bump.LineupToChargeThresh and self.target_line().delta().dot(self.robot.pos - main.ball().pos) <= -constants.Robot.Radius and not self.facing_err_above_threshold(),
            'lined up')

        # FIXME: this condition was never setup in the C++ one...
        self.add_transition(
            Bump.State.charge, behavior.Behavior.State.completed,
            lambda: (main.ball().pos - self.robot.pos).mag() < (constants.Robot.Radius + constants.Ball.Radius + 0.03),
            'ball has been bumped')

        self.add_transition(
            Bump.State.charge, Bump.State.lineup,
            lambda: robocup.Line(self.robot.pos, self.target).dist_to(main.ball().pos) > Bump.EscapeChargeThresh,
            'bad ball placement')

    ## the Point we're trying to bump the ball towards
    @property
    def target(self):
        return self._target

    @target.setter
    def target(self, value):
        self._target = value

    def facing_err_above_threshold(self):
        dirVec = robocup.Point(
            math.cos(self.robot.angle), math.sin(self.robot.angle))
        facing_thresh = math.cos(Bump.FacingThresh)
        facing_err = dirVec.dot((self.target - main.ball().pos).normalized())
        # NOTE: yes, the comparator is backwards, that's the way it was in the c++ one...
        # TODO: remove the above note and clarify what's going on
        return facing_err < facing_thresh

    # line from ball to target
    def target_line(self):
        return robocup.Line(main.ball().pos, self.target)

    def execute_lineup(self):
        target_line = self.target_line()
        target_dir = target_line.delta().normalized()
        behind_line = robocup.Segment(
            main.ball().pos - target_dir *
            (Bump.DriveAroundDist + constants.Robot.Radius),
            main.ball().pos - target_dir * 5.0)
        if target_line.delta().dot(self.robot.pos - main.ball(
        ).pos) > -constants.Robot.Radius:
            # we're very close to or in front of the ball
            self.robot.set_avoid_ball_radius(Bump.LineupBallAvoidRadius)
            self.robot.move_to(main.ball().pos - target_dir * (
                Bump.DriveAroundDist + constants.Robot.Radius))
        else:
            self.robot.set_avoid_ball_radius(Bump.LineupBallAvoidRadius)
            self.robot.move_to(behind_line.nearest_point(self.robot.pos))
            main.system_state().draw_line(behind_line, constants.Colors.Black,
                                          "Bump")

        delta_facing = self.target - main.ball().pos
        self.robot.face(self.robot.pos + delta_facing)

    def execute_charge(self):
        main.system_state().draw_line(
            robocup.Line(self.robot.pos, self.target), constants.Colors.White,
            "bump")
        main.system_state().draw_line(
            robocup.Line(main.ball().pos, self.target), constants.Colors.White,
            "bump")

        ball2target = (self.target - main.ball().pos).normalized()
        drive_dir = (main.ball().pos - ball2target *
                     constants.Robot.Radius) - self.robot.pos

        # we want to drive toward the ball without using the path planner
        # we do this by setting the speed directly
        # AccelBias forces us to accelerate a bit more
        speed = self.robot.vel.mag() + Bump.AccelBias

        self.robot.set_world_vel(drive_dir.normalized() * speed)
        self.robot.face(main.ball().pos)
